File: dart/lib/twocrystal_balls.dart
import 'dart:math';

int twoCrystalBalls(List<bool> breaks) {
  var jmpAmount = sqrt(breaks.length).floor();

  var i = jmpAmount;
  for (; i < breaks.length; i += jmpAmount) {
    if (breaks[i]) {
      break;
    }
  }

  i -= jmpAmount;

  for (var j = 0; j <= jmpAmount && i < breaks.length; ++j, ++i) {
    if (breaks[i]) {
      return i;
    }
  }

  return -1;
}

File: kotlin/algorithms/src/main/kotlin/searching/TwoCrystalBalls.kt
package searching

import kotlin.math.sqrt

fun twoCrystalBalls(breaks: BooleanArray): Int {
    val jmpAmount = sqrt(breaks.size.toDouble()).toInt();

    var i = jmpAmount;
    while (i < breaks.size) {
        if (breaks[i]) {
             break;
        }

        i += jmpAmount;
    }

    i -= jmpAmount
    for (j in 0..jmpAmount) {
        if (i < breaks.size && breaks[i]) {
            return i
        }
        i++
    }

    return -1;
}